### ISSUE TITLE
kubectl cp to work on stopped/completed pods

### ISSUE THREAD

**Issue by xellsys** [kubernetes/kubectl#454]
**Status Quo:**
The current implementation of `kubectl cp` requires the container to be running by encapsulating the exec command using the tar binary.

**Requirement:**
Allow copying files _from_ (possibly also _to_) a stopped container.

**Additional info:**
This relates to #58512.
It would also more closely align to the `docker cp` functionality.
No need for the tar binary in the container anymore.

-----
**Background:**
My current use case is running `helm test` with a more sophisticated test set (0 and 1 results do not suffice for analysis) and having a _simple_ way of persisting test results at the end (Jenkins or w/e).
I know there are other solutions but I want to keep my test pod simple (not actively pushing test results to some endpoint or keeping it alive) and would like to avoid extensive configuration of a persistent storage.

**Comment by tsuna** [comment]
Would be great to have this capability.

**Comment by tcoopman** [comment]
Ran into the same issue, but we didn't have to transfer a lot of data. The easiest **hack** was to tar all the data and base64 encode is and log it to stdout. Then you can get the data out because you can still get the logs for a stopped container.

**Comment by zedtux** [comment]
I'm also looking to this feature !

**Comment by rafaelaazevedo** [comment]
this would be great for people who run tests on a kubernetes pod and need to get the reports after

**Comment by wjcheng1** [comment]
yes would love to have this ability!

**Comment by yue9944882** [comment]
ref [https://github.com/kubernetes/kubernetes/issues/58512](https://github.com/kubernetes/kubernetes/issues/58512)

**Comment by stmaute** [comment]
I also would like to have this feature. Are there any plans to implement this???

**Comment by cliffburdick** [comment]
This would be really useful as large files base64 encoded is not manageable.

**Comment by seans3** [comment]
I'm not currently sure how feasible this is, but there seems to be a lot of interest--raising the priority.

/priority P2

**Comment by FilipB** [comment]
Not sure if it's applicable here, but I was advised to use nice workaround for OpenShift 4.1.
container_ID is needed.
Then you can run on your node where the container was running: 
sed -n  '/path/ s/merged/diff\//p' /var/lib/containers/storage/overlay-containers/$container_ID/userdata/config.json
which will give you location of directory which contains added files to your container e.g.:
/var/lib/containers/storage/overlay/406944605cabd99dc6e886ab9c3998b89fb25866e135c0906f573da47ffe23e1/diff/

That worked for me to get created files from exited container.

**Comment by rocketraman** [comment]
Another use case, slightly different than obtaining test results mentioned a couple of times above: containers that dump core or heap will obviously terminate after doing so. At this point, doing `kubectl cp` on a stopped container is necessary to obtain the core/heap. One example is using the `-XX:+HeapDumpOnOutOfMemoryError` JVM flag.

**Comment by jglick** [comment]
A related use case: copying files to or from the persistent volume associated with a scaled-down stateful set, without having to scale it up or attach the volume to a temporary no-op pod.

**Comment by samsieber** [comment]
So I too am working with k8s, triggering jobs from jenkins and wanting files back. the method I've settled on is a sidecar that opens a fifo called `export-files-fifo` (via the `mkfifo` command) at the beginning and in the `preStop` livecycle hook tars and pipes all the files I'm interested over the fifo. And the jenkins job triggering the job immediately run something along the lines of `kubectl exec my-job-pod -- sh 'cat | export-files-fifo' > exported_files.tar`

**Comment by atheiman** [comment]
@samsieber could you provide a simple yaml example of this?

**Comment by samsieber** [comment]
@atheiman Here it is: https://gist.github.com/samsieber/349fde899d508b4e6be119e762fb600c - the job stays active until the './grab-files.sh' is called, at which point the job shuts down automatically.

**Comment by j-mew-s** [comment]
I'm debugging an error in file modification carried out by a Dockerfile. It is modifying a config file, and the error prevents the container from starting (CrashLoopBackOff). I cannot inspect it purely by checking out the image and running with a different command since it uses environment variables set at docker run time.

This feature would be helpful for me to quickly understand the state of the config file and identify the error. Instead, I'm going to log the file's contents temporarily in the docker file, then revert once I understand the issue (which will involve 2 build/test cycles).

**Comment by bolt-juri-gavshin** [comment]
Hello! Is there any ETA on fixing this?
It is really inconvenient to have so complex workarounds just to get files from "Completed" pods of a job. Have to create pods directly with init containers and do some magic just do such a simple operation...
Thanks a lot in advance!

**Comment by brianpursley** [comment]
Just to understand the issue a little better (correct me if I'm wrong)...

You have something like this...
```
kubectl run job1 --image=busybox --restart=Never -- /bin/sh -c "echo test > /output.txt"
```

And you want to be able to do this after the job has completed...
```
kubectl cp job1:/output.txt output.txt
```

But it fails with this message:
```
error: cannot exec into a container in a completed pod; current phase is Succeeded
```

The reason you don't want to define a persistent volume is that you just want to get the result of your job this one time, it is not something that is needed long-term, is that correct?

**Comment by bolt-juri-gavshin** [comment]
@brianpursley , you are 100% correct.
At the moment what I do is create a pod with `tail -f /dev/null` as command and then use `kubectl exec` + `kubectl cp` + `kubectl delete`.
One of use cases is execution of tests in Jenkins, after execution I need to retrieve junit.xml file.
I have ok-ish workaround now (with just 1 additional step), but being able to copy files from Completed pod would require 1 step less...

**Comment by alecor191** [comment]
@bolt-juri-gavshin I have a similar use case as you describe (exec pod that runs tests, then pull the test logs from the pod). Do you mind sharing more details on how your approach works? Specifically, I understand that you keep the "test pod" alive with `tail -f /dev/null`, however, how do you know when the test run is complete, so that you can copy the test result files with `kubectl cp`?

**Comment by gionapaolini** [comment]
> 
> 
> @brianpursley , you are 100% correct.
> At the moment what I do is create a pod with `tail -f /dev/null` as command and then use `kubectl exec` + `kubectl cp` + `kubectl delete`.
> One of use cases is execution of tests in Jenkins, after execution I need to retrieve junit.xml file.
> I have ok-ish workaround now (with just 1 additional step), but being able to copy files from Completed pod would require 1 step less...

Can you elaborate a bit more on your workaround?
That's how I understood:

1. Build an image for your test project and use tail -f /dev/null as entrypoint
2. Run a pod with that image
3. Execute a shell into it and run your tests
4. Copy the reports
5. Delete the pod

Have I understood correctly?

**Comment by lindhe** [comment]
I would be interested in trying to resolve this issue. But I realize that this is an intricate problem which may require some expertise from more experienced Kubernetes developers. Are there any suggestions on reasonable designs to approach this, or do we know of something that makes this problem intrinsically hard to solve?

**Comment by AshwinNS** [comment]
> > @brianpursley , you are 100% correct.
> > At the moment what I do is create a pod with `tail -f /dev/null` as command and then use `kubectl exec` + `kubectl cp` + `kubectl delete`.
> > One of use cases is execution of tests in Jenkins, after execution I need to retrieve junit.xml file.
> > I have ok-ish workaround now (with just 1 additional step), but being able to copy files from Completed pod would require 1 step less...
> 
> Can you elaborate a bit more on your workaround?
> That's how I understood:
> 
> 1. Build an image for your test project and use tail -f /dev/null as entrypoint
> 2. Run a pod with that image
> 3. Execute a shell into it and run your tests
> 4. Copy the reports
> 5. Delete the pod
> 
> Have I understood correctly?

sounds good, you can even override the `entrypoint` to sleep the container for certain amount of time, that would be more elegant.

**Comment by cassador** [comment]
Any update or workaround on this?
I saw the mentioned workaround but i don't think they could be used in our configuraiton.
We are running it on EKS cluster with SSH disabled so i can't use that one with where the pods data are being stored by default on runtime,

**Comment by pisto** [comment]
Please also make this work on pods in CrashLoopBackoff state. It's painful to debug these containers when you suspect the filesystem in the pod is incorrect, but there's no way to cp files, or even just list the container filesystem when a container is not running, and workarounds may be difficutl to apply.

**Comment by rptaylor** [comment]
I agree this would be useful. People need a way to extract files produced in pods after they finish.

My use case is a job that requires a lot of resources and runs for a couple days. I don't want to keep it running longer than necessary with e.g. 'sleep' because that would waste resources, but I also _really_ don't want to lose the output if I am too busy to check on it and retrieve the output until several days afterwards. The output is important but reasonably small, only ~< 1 MB of text files.  I could write the output to a persistent volume but that only helps slightly because it is not very conveniently accessible -  I would have to start up another pod just for data extraction, to mount the volume and then 'kubectl cp'.

kubectl cp from a finished pod would be one way, but what if you could do something like `cp  mydata  /dev/kubebucket` as the last step of execution in your pod before it finishes, where /dev/kubebucket is a device exposed by kubelet to your pod that you can write to, for reasonably small and short-term storage of pod outputs, which is retained for a period of time after pod completion. Then suppose you could `kubectl mount <pod or bucket identifier>  <optional -c container>  mountdir`  to get a read-only FUSE mount of whatever you put in the "bucket" , so you can get your files out and then `fusermount -u mountdir` (or `kubectl unmount mountdir`).  Perhaps you could also `kubectl get bucket <bucketID>`  , `kubectl delete bucket <bucketID>` etc.
There could be configurable max size limits and automatic expiration time periods to make sure the buckets don't accumulate too much over time.

Just an idea, no idea how feasible to implement it would be but it could be an alternative way to extract outputs from completed containers by providing some short term storage facility that is outside of the pods.

**Comment by Alexander-Bartosh** [comment]
I know that this is not a lot of people desire, but as a workaround you can have a sidecar (keep-alive sleep ) container running with shareProcessNamespace: true. 
Both should share a volume for results that you want to copy out. 
Once the tests container completes: 
kubectl logs -f -n "$namespace" -c "$podcontainer" "$pod"  
exist you can pick up you test results from the share volume using kubectl cp and stop/remove the pod/job

**Comment by nagakonduri** [comment]
Looking forward for this feature, please share if this is in consideration.

**Comment by rafaelaazevedo** [comment]
I have been using this workaround, not ideal, but it is working.

[How to get your test reports from a pod – Kubernetes](https://azevedorafaela.com/2019/03/10/how-to-get-your-test-reports-from-a-pod-kubernetes/)

**Comment by tclarke** [comment]
I'd like to see this too. My use case is for post mortem debugging of production pods. (so an extrypoint of "sleep 10000" isn't an option)

**Comment by Warxcell** [comment]
+1. Useful to run unit tests in pod and get coverage reports, which are generated after tests finishes.

**Comment by graydon** [comment]
This (and in general "access to the pod filesystem using an API that is not just executing-tar-on-stdin") would be very desirable to me. I am sort of astonished that this is not part of the standard k8s API. Files in the filesystem -- not just open filehandles -- are actually important parts of the lifecycle and system interface of most programs.

**Comment by brianpursley** [comment]
I still think using a persistent volume is the best way to preserve and access the output produced by a completed pod

But I know there are a lot of people who end up here, wanting a way to run some task to completion and then use `kubectl cp` to access the output that was produced inside one of the container's filesystem.

For anyone who doesn't want to create a persistent volume, for whatever reason, here is an idea I came up with that might work for you.

It makes use of:
* A **worker** container that is actually an init container.  This is what runs automated tests or whatever other task you want to run in the pod.
* A **keepalive** container that does nothing other than keep the pod alive so you can copy the output
* A **results** ephemeral volume (emptyDir) to share the output of the worker with the keepalive container

The pod will not become "ready" until the init container completes its work.  Because of this, you can use `kubectl wait` to wait for the worker to finish and the pod to become ready.

Once the pod is ready, the keepalive container keeps the pod alive, so you have an opportunity (60 seconds in my example) to use `kubectl cp` to get the files stored in the results volume.

```
kubectl apply -f - <<EOF
apiVersion: v1
kind: Pod
metadata:
  name: test
spec:
  volumes:
  - name: results
    emptyDir: {}
  initContainers:
  - image: busybox
    name: worker
    command: ["/bin/sh", "-c", "echo 'hello from the worker' > /results/output.txt;"]
    volumeMounts:
    - name: results
      mountPath: /results
  containers:
  - image: busybox
    name: keepalive
    command: ["/bin/sh", "-c", "sleep 60"]
    volumeMounts:
    - name: results
      mountPath: /results
EOF

kubectl wait --timeout=1h --for=condition=Ready=true pod/test
kubectl cp test:/results/output.txt /tmp/output.txt
kubectl delete pod test
cat /tmp/output.txt
```

**Comment by Warxcell** [comment]
> I still think using a persistent volume is the best way to preserve and access the output produced by a completed pod
> 
> But I know there are a lot of people who end up here, wanting a way to run some task to completion and then use `kubectl cp` to access the output that was produced inside one of the container's filesystem.
> 
> For anyone who doesn't want to create a persistent volume, for whatever reason, here is an idea I came up with that might work for you.
> 
> It makes use of:
> 
> * A **worker** container that is actually an init container.  This is what runs automated tests or whatever other task you want to run in the pod.
> * A **keepalive** container that does nothing other than keep the pod alive so you can copy the output
> * A **results** ephemeral volume (emptyDir) to share the output of the worker with the keepalive container
> 
> The pod will not become "ready" until the init container completes its work. Because of this, you can use `kubectl wait` to wait for the worker to finish and the pod to become ready.
> 
> Once the pod is ready, the keepalive container keeps the pod alive, so you have an opportunity (60 seconds in my example) to use `kubectl cp` to get the files stored in the results volume.
> 
> ```
> kubectl apply -f - <<EOF
> apiVersion: v1
> kind: Pod
> metadata:
>   name: test
> spec:
>   volumes:
>   - name: results
>     emptyDir: {}
>   initContainers:
>   - image: busybox
>     name: worker
>     command: ["/bin/sh", "-c", "echo 'hello from the worker' > /results/output.txt;"]
>     volumeMounts:
>     - name: results
>       mountPath: /results
>   containers:
>   - image: busybox
>     name: keepalive
>     command: ["/bin/sh", "-c", "sleep 60"]
>     volumeMounts:
>     - name: results
>       mountPath: /results
> EOF
> 
> kubectl wait --timeout=1h --for=condition=Ready=true pod/test
> kubectl cp test:/results/output.txt /tmp/output.txt
> kubectl delete pod test
> cat /tmp/output.txt
> ```

Problem with this approach is that `kubectl wait --timeout=1h --for=condition=Ready=true pod/test` doesn't return exit code of init container.

**Comment by MikaelSmith** [comment]
I've put together a helper that wraps up a workflow for this problem: https://github.com/puppetlabs/kubectl-ran

It's reasonably complete, although I plan to add an ability to customize the pod spec a bit to add sidecars.

**Comment by vishrantgupta** [comment]
You can add another container that will run infinitely with the restart policy `Never`

Example:

```
apiVersion: v1
kind: Pod
metadata:
  labels:
    app: test-automation
    instance: testautomation-sample
  name: testautomation-sample
  namespace: default
spec:
  containers:
  - args:
    - make
    - component_test
    image: test-automation:latest
    imagePullPolicy: Always
    name: component-test
    resources: {}
    securityContext:
      runAsUser: 0
  - args:
    - /usr/bin/tail
    - -f
    - /dev/null
    image: test-automation:latest
    imagePullPolicy: IfNotPresent
    name: wait-for-report
    resources: {}
  imagePullSecrets:
  - name: kubelet-pull-secret
  restartPolicy: Never
  securityContext:
    runAsUser: 0
  serviceAccount: default
  serviceAccountName: default
```

This will give you control over running `kubectl cp ...` followed by `kubectl delete ...`

**Comment by HHobeck** [comment]
Dear community.

Almost four years later this topic is still open. I’m really afraid that such thinks like ergonomic and API friendly usage of Kubernetes are not mentioned with the attention they should be. Of course there are workarounds to solve this problem. But this workarounds are really hard to implement and error prone. From the architecture point of view it’s also sub-optimal to have circular references or use shared file system in some cases. Putting the main function into a init routine is a absolute no-go.

Why it is permitted to access the container after the job finished or after a pod failed to start!? It’s not only accessing the data. It’s also necessary for analyzing purpose to attach the container.

**Please increase the priority.**

Thank you very much.

**Comment by seans3** [comment]
This functionality (keeping a pod alive after completion) is beyond the scope of `kubectl`. Suggest closing this issue. If it is addressed, then it needs to be addressed more holistically including other areas (e.g. `kubelet`).

**Comment by brianpursley** [comment]
The use case is definitely understandable: wanting to run something in a pod, and after it is finished to retrieve the output of what was run.  

The reason this issue has been open since 2018 is that this isn't functionality that kubectl is able to provide without changes first being made deeper in the Kubernetes stack (kubelet, etc).

There are some options for you today though:  

1. Save you pod's output to a persistent volume. This is the recommended way to do it.

2. If you don't want to use a PV for some reason, there are some workarounds in the comments above involving multi-container pods.

**However, If you still feel very strongly about wanting this particular feature**, to be able to copy files from stopped pods, then my suggestion is to open an issue on the [kubernetes/kubernetes](https://github.com/kubernetes/kubernetes/issues/new?assignees=&labels=kind%2Ffeature&template=enhancement.yaml) repo where it can be addressed by another SIG who is able to make the necessary changes to support this.

I realize this is a popular issue, but I'm going to go ahead and close it based on what I said above.  If Kubernetes is changed in the future to support this server-side, then we can make the necessary changes in kubectl to provide this functionality.

/close

**Comment by HHobeck** [comment]
Hi there.
I have created a new issue: https://github.com/kubernetes/kubernetes/issues/111045
Please support me and give this issue a weight. Not less important please help me to find the persons who are able to bring this topic to success.

**Comment by AymenFJA** [comment]
any updates on this ticket?

**Comment by HHobeck** [comment]
> any updates on this ticket?

Nope. I would say there is zero progress.
